app/models/canonical/sync/association_syncer.rb

```diff
@@ -0,0 +1,76 @@
+# frozen_string_literal: true
+
+module Canonical
+  module Sync
+    class AssociationSyncer < Syncer
```

Comment:
The code in this class is ugly and has major metaprogramming energy, however, because of the complexity, generalising and extracting it seemed like the lesser evil compared to having similarly complex but less metaprogrammy code spread out across several classes.